### PR TITLE
8334026: Provide a diagnostic PrintMemoryMapAtExit switch on Linux

### PR DESCRIPTION
**Pull request by tstuefe**
To investigate footprint problems of processes with a short lifespan (e.g. startup footprint problems), we found it very useful to be able to get an annotated memory map at exit - the same map printed by jcmd System.map.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334026](https://bugs.openjdk.org/browse/JDK-8334026): Provide a diagnostic PrintMemoryMapAtExit switch on Linux (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Matthias Baesken](https://openjdk.org/census#mbaesken) (@MBaesken - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19660/head:pull/19660` \
`$ git checkout pull/19660`

Update a local copy of the PR: \
`$ git checkout pull/19660` \
`$ git pull https://git.openjdk.org/jdk.git pull/19660/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19660`

View PR using the GUI difftool: \
`$ git pr show -t 19660`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19660.diff">https://git.openjdk.org/jdk/pull/19660.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19660#issuecomment-2162278928)

### CODE REVIEW

File: test/hotspot/jtreg/runtime/NMT/PrintMemoryMapAtExitTest.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import jdk.test.lib.process.OutputAnalyzer;
+import jdk.test.lib.process.ProcessTools;
+
+public class PrintMemoryMapAtExitTest {
+
+    public static void main(String args[]) throws Exception {
+
+        ProcessBuilder pb = ProcessTools.createTestJavaProcessBuilder(
+      "-XX:+UnlockDiagnosticVMOptions", "-XX:+PrintMemoryMapAtExit",
+                "-XX:NativeMemoryTracking=summary", "-version");
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/NMT/PrintMemoryMapAtExitTest.java:43]: Nit: indentation is off here.

File: test/hotspot/jtreg/runtime/NMT/PrintMemoryMapAtExitTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.shouldContain("Memory mappings");
+        output.shouldContain("JAVAHEAP");
+        output.shouldHaveExitValue(0);
```
Comment (dholmes-ora) [inline test/hotspot/jtreg/runtime/NMT/PrintMemoryMapAtExitTest.java:48]: Suggestion: check the exit value first. If we crashed then the other checks will likely fail before we get here.

File: src/hotspot/share/nmt/memMapPrinter.cpp [hunk 3]
```diff
@@ -203,6 +204,8 @@ static void print_thread_details(uintx thread_id, const char* name, outputStream
 // Given a region [from, to), if it intersects a known thread stack, print detail infos about that thread.
 static void print_thread_details_for_supposed_stack_address(const void* from, const void* to, outputStream* st) {
 
+  ResourceMark rm;
```
Comment (dholmes-ora) [inline src/hotspot/share/nmt/memMapPrinter.cpp:207]: Why was this needed to be added here? Shouldn't callers have any needed ResourceMarks?

**Review by dholmes-ora — COMMENTED** [review]
Seems quite reasonable. One query and a couple of minor nits.

Thanks

**Comment by tstuefe** [comment]
Thanks @dholmes-ora !

I worked in your remarks.

File: src/hotspot/share/nmt/memMapPrinter.cpp (hunk 3 above)
Comment (dholmes-ora) [reply src/hotspot/share/nmt/memMapPrinter.cpp:207]: Okay. I was trying to find where the existing callers have their RM and it is not obvious.

**Comment by tstuefe** [comment]
Thanks @MBaesken and @dholmes-ora !

/integrate

File: src/hotspot/share/nmt/memMapPrinter.cpp (hunk 3 above)
Comment (tstuefe) [reply src/hotspot/share/nmt/memMapPrinter.cpp:207]: Existing caller is DCmd::parse_and_execute, where we have an RM that guards every individual Dcmd execution.
